File: plugins/capacitor-gr-mjpeg/ios/Sources/GrMjpegPlugin/GrMjpegPlugin.swift
import Foundation
import Capacitor
import UIKit

/// Native MJPEG viewer drawn UNDER a transparent WKWebView, so HTML overlays (HUD,
/// tap-to-focus) stay on top and interactive. Frames never cross the JS bridge.
@objc(GrMjpegPlugin)
public class GrMjpegPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "GrMjpegPlugin"
    public let jsName = "GrMjpeg"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "start", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setRect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setUrl", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pause", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resume", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stop", returnType: CAPPluginReturnPromise)
    ]

    private var container: UIView?
    private var imageView: UIImageView?
    private var streamer: MjpegStreamer?
    private var pausedByJs = false

    override public func load() {
        let nc = NotificationCenter.default
        nc.addObserver(self, selector: #selector(appDidEnterBackground),
                       name: UIApplication.didEnterBackgroundNotification, object: nil)
        nc.addObserver(self, selector: #selector(appWillEnterForeground),
                       name: UIApplication.willEnterForegroundNotification, object: nil)
    }

    deinit { NotificationCenter.default.removeObserver(self) }

    @objc private func appDidEnterBackground() { streamer?.pause() }
    @objc private func appWillEnterForeground() { if !pausedByJs { streamer?.resume() } }

    @objc func start(_ call: CAPPluginCall) {
        guard let urlString = call.getString("url"), let url = URL(string: urlString),
              let rect = call.getObject("rect") else {
            call.reject("url and rect are required")
            return
        }
        let stall = TimeInterval(call.getInt("stallTimeoutMs") ?? 4000) / 1000
        let background = Self.color(call.getString("backgroundColor") ?? "#000000")
        var headers: [String: String] = [:]
        for (k, v) in call.getObject("headers") ?? [:] {
            if let s = v as? String { headers[k] = s }
        }

        DispatchQueue.main.async {
            self.teardown()
            guard let webView = self.webView, let host = webView.superview else {
                call.reject("No web view")
                return
            }

            let container = UIView(frame: host.bounds)
            container.autoresizingMask = [.flexibleWidth, .flexibleHeight]
            container.backgroundColor = background
            container.isUserInteractionEnabled = false
            host.insertSubview(container, belowSubview: webView)

            let imageView = UIImageView()
            imageView.contentMode = .scaleAspectFit
            imageView.backgroundColor = .black
            container.addSubview(imageView)
            self.container = container
            self.imageView = imageView
            self.apply(rect: rect)

            // Let the native view show through the page.
            webView.isOpaque = false
            webView.backgroundColor = .clear
            webView.scrollView.backgroundColor = .clear

            let streamer = MjpegStreamer(url: url, headers: headers, stall: stall, imageView: imageView)
            streamer.onState = { [weak self] state, message in
                var e: JSObject = ["state": state]
                if let message = message { e["message"] = message }
                self?.notifyListeners("state", data: e)
            }
            streamer.onStats = { [weak self] stats in
                var e: JSObject = [:]
                for (k, v) in stats {
                    if let n = v as? Int { e[k] = n } else if let s = v as? String { e[k] = s }
                }
                self?.notifyListeners("stats", data: e)
            }
            self.pausedByJs = false
            self.streamer = streamer
            streamer.start()
            call.resolve()
        }
    }

    @objc func setRect(_ call: CAPPluginCall) {
        let rect = call.options as? [String: Any] ?? [:]
        DispatchQueue.main.async {
            self.apply(rect: rect)
            call.resolve()
        }
    }

    @objc func setUrl(_ call: CAPPluginCall) {
        guard let s = call.getString("url"), let url = URL(string: s) else {
            call.reject("url is required")
            return
        }
        streamer?.setUrl(url)
        call.resolve()
    }

    @objc func pause(_ call: CAPPluginCall) {
        pausedByJs = true
        streamer?.pause()
        call.resolve()
    }

    @objc func resume(_ call: CAPPluginCall) {
        pausedByJs = false
        streamer?.resume()
        call.resolve()
    }

    @objc func stop(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.teardown()
            call.resolve()
        }
    }

    // MARK: - Helpers

    /// CSS px (relative to the web view's viewport) == points; offset by the web view's frame.
    private func apply(rect: [String: Any]) {
        guard let imageView = imageView, let webView = webView else { return }
        func num(_ k: String) -> CGFloat { CGFloat((rect[k] as? NSNumber)?.doubleValue ?? 0) }
        imageView.frame = CGRect(x: webView.frame.minX + num("x"),
                                 y: webView.frame.minY + num("y"),
                                 width: max(0, num("width")),
                                 height: max(0, num("height")))
    }

    private func teardown() {
        streamer?.stop()
        streamer = nil
        container?.removeFromSuperview()
        container = nil
        imageView = nil
    }

    private static func color(_ hex: String) -> UIColor {
        var s = hex.trimmingCharacters(in: .whitespaces)
        if s.hasPrefix("#") { s.removeFirst() }
        guard s.count == 6 || s.count == 8, let v = UInt64(s, radix: 16) else { return .black }
        let hasAlpha = s.count == 8
        let r = CGFloat((v >> (hasAlpha ? 24 : 16)) & 0xFF) / 255
        let g = CGFloat((v >> (hasAlpha ? 16 : 8)) & 0xFF) / 255
        let b = CGFloat((v >> (hasAlpha ? 8 : 0)) & 0xFF) / 255
        let a = hasAlpha ? CGFloat(v & 0xFF) / 255 : 1
        return UIColor(red: r, green: g, blue: b, alpha: a)
    }
}

File: plugins/capacitor-gr-mjpeg/android/src/main/java/com/grremote/plugins/grmjpeg/GrMjpegPlugin.java
package com.grremote.plugins.grmjpeg;

import android.graphics.Color;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebView;
import android.widget.FrameLayout;
import android.widget.ImageView;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * Native MJPEG viewer drawn UNDER a transparent WebView, so HTML overlays (HUD, tap-to-focus)
 * stay on top and interactive. Frames never cross the JS bridge.
 */
@CapacitorPlugin(name = "GrMjpeg")
public class GrMjpegPlugin extends Plugin {

    private FrameLayout container;
    private ImageView imageView;
    private MjpegStreamer streamer;
    private boolean pausedByJs = false;

    @PluginMethod
    public void start(PluginCall call) {
        final String url = call.getString("url");
        final JSObject rect = call.getObject("rect");
        if (url == null || rect == null) {
            call.reject("url and rect are required");
            return;
        }
        final int stallMs = call.getInt("stallTimeoutMs", 4000);
        final int bg = parseColor(call.getString("backgroundColor", "#000000"));
        final Map<String, String> headers = toMap(call.getObject("headers", new JSObject()));

        getBridge().executeOnMainThread(() -> {
            teardown();
            WebView webView = getBridge().getWebView();
            ViewGroup parent = (ViewGroup) webView.getParent();

            container = new FrameLayout(getContext());
            container.setBackgroundColor(bg);
            container.setClickable(false);
            container.setImportantForAccessibility(View.IMPORTANT_FOR_ACCESSIBILITY_NO);
            parent.addView(container, parent.indexOfChild(webView),
                new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));

            imageView = new ImageView(getContext());
            imageView.setScaleType(ImageView.ScaleType.FIT_CENTER);
            imageView.setBackgroundColor(Color.BLACK);
            container.addView(imageView, new FrameLayout.LayoutParams(0, 0));
            applyRect(rect);

            // Let the native view show through the page.
            webView.setBackgroundColor(Color.TRANSPARENT);

            pausedByJs = false;
            streamer = new MjpegStreamer(url, headers, stallMs, imageView, new MjpegStreamer.Listener() {
                @Override
                public void onState(String state, String message) {
                    JSObject e = new JSObject();
                    e.put("state", state);
                    if (message != null) e.put("message", message);
                    notifyListeners("state", e);
                }

                @Override
                public void onStats(int fps, int dropped, int width, int height, String contentType, String mode) {
                    JSObject e = new JSObject();
                    e.put("fps", fps);
                    e.put("dropped", dropped);
                    e.put("width", width);
                    e.put("height", height);
                    e.put("contentType", contentType);
                    e.put("mode", mode);
                    notifyListeners("stats", e);
                }
            });
            streamer.start();
            call.resolve();
        });
    }

    @PluginMethod
    public void setRect(PluginCall call) {
        final JSObject rect = call.getData();
        getBridge().executeOnMainThread(() -> {
            applyRect(rect);
            call.resolve();
        });
    }

    @PluginMethod
    public void setUrl(PluginCall call) {
        String url = call.getString("url");
        if (url == null) {
            call.reject("url is required");
            return;
        }
        if (streamer != null) streamer.setUrl(url);
        call.resolve();
    }

    @PluginMethod
    public void pause(PluginCall call) {
        pausedByJs = true;
        if (streamer != null) streamer.pause();
        call.resolve();
    }

    @PluginMethod
    public void resume(PluginCall call) {
        pausedByJs = false;
        if (streamer != null) streamer.resume();
        call.resolve();
    }

    @PluginMethod
    public void stop(PluginCall call) {
        getBridge().executeOnMainThread(() -> {
            teardown();
            call.resolve();
        });
    }

    // App lifecycle: never stream in the background.
    @Override
    protected void handleOnPause() {
        if (streamer != null) streamer.pause();
    }

    @Override
    protected void handleOnResume() {
        if (streamer != null && !pausedByJs) streamer.resume();
    }

    @Override
    protected void handleOnDestroy() {
        if (streamer != null) streamer.stop();
        streamer = null;
    }

    // ───────────── helpers ─────────────

    /** CSS px (relative to the WebView viewport) → device px in the container. */
    private void applyRect(JSObject rect) {
        if (imageView == null || rect == null) return;
        WebView webView = getBridge().getWebView();
        float d = getContext().getResources().getDisplayMetrics().density;
        FrameLayout.LayoutParams lp = new FrameLayout.LayoutParams(
            Math.max(0, Math.round((float) rect.optDouble("width", 0) * d)),
            Math.max(0, Math.round((float) rect.optDouble("height", 0) * d)));
        lp.leftMargin = webView.getLeft() + Math.round((float) rect.optDouble("x", 0) * d);
        lp.topMargin = webView.getTop() + Math.round((float) rect.optDouble("y", 0) * d);
        imageView.setLayoutParams(lp);
    }

    private void teardown() {
        if (streamer != null) {
            streamer.stop();
            streamer = null;
        }
        if (container != null) {
            ViewGroup parent = (ViewGroup) container.getParent();
            if (parent != null) parent.removeView(container);
            container = null;
            imageView = null;
        }
    }

    private static Map<String, String> toMap(JSObject obj) {
        Map<String, String> map = new HashMap<>();
        Iterator<String> keys = obj.keys();
        while (keys.hasNext()) {
            String k = keys.next();
            String v = obj.optString(k, null);
            if (v != null) map.put(k, v);
        }
        return map;
    }

    private static int parseColor(String hex) {
        try {
            return Color.parseColor(hex);
        } catch (Exception e) {
            return Color.BLACK;
        }
    }
}
